### PR TITLE
added function parameter for secure cookies, added test, added api ex…

### PR DESCRIPTION
**Pull request by ry8806**
added function parameter in localstorageservice.cookie.set() for secure cookies client side using the "secure" attribute mentioned in the W3C CSP Headers Spec (server side): https://www.w3.org/TR/csp-cookies/#grammardef-secure

added tests and fixed some of the README.md

### CODE REVIEW

**Comment by ry8806** [comment]
#272

**Comment by ealves-pt** [comment]
@ry8806 what about the config phase? For example, If by default I want to set all the cookies to be stored as secure?

**Comment by ealves-pt** [comment]
This PR will close this old PR, #185

**Comment by ealves-pt** [comment]
@ry8806 when you can rebase with the master and update the PR.

**Comment by ealves-pt** [comment]
This PR also closes #250

**Comment by ry8806** [comment]
@ealves-pt I'll get on this and implement the config phase, I'll aim to do it within 24hrs

**Comment by ealves-pt** [comment]
:+1:

**Comment by ry8806** [comment]
hi, I've updated it to work with config (and updated the README and added more tests) - had a bit of trouble with Git so let me know if I've rebased correctly - hopefully it should all be there!

File: src/angular-local-storage.js [hunk 1]
```diff
@@ -311,9 +313,23 @@ angular
             }
             if (!!key) {
               var cookiePath = "; path=" + cookie.path;
-              if(cookie.domain){
+              if (cookie.domain) {
                 cookieDomain = "; domain=" + cookie.domain;
               }
+              /* Providing the secure parameter always takes precedence over config 
+               * (allows developer to mix and match secure + non-secure) */
+              if (isDefined(secure) && secure !== null) {
```
Comment (ealves-pt) [inline src/angular-local-storage.js]: Why not replace this nested `ifs` for:
```javascript
if (typeof secure === 'boolean' && secure) {
  cookieDomain += '; secure';
}
```

I think it makes the code more readable.

**Comment by ealves-pt** [comment]
Nice PR. Just commented one thing and I will merge it after your feedback.

Thanks for the work :+1:

File: src/angular-local-storage.js (hunk 1 above)
Comment (ry8806) [inline src/angular-local-storage.js]: yeh, good catch!
thanks

Comment (ealves-pt) [inline src/angular-local-storage.js]: Any update regarding this? Will you do the change or should I merge it as is?

Comment (ry8806) [inline src/angular-local-storage.js]: If I take the code above, we lose the functionality of being able to explicitly override what config says.

At the moment - if config specifies `secure = true` I can call `cookie.set("key", "value", 3, false)` and it allows me to override what the config says (by going into the first "if" and obviously not going into the `if-else`).

If I take the code above and we specify secure as `false` we lose that functionality and will fall in to the `if-else` which will take the value from config. Which renders the secure parameter on `cookie.set` useless...

what do you think???

**Comment by ry8806** [comment]
I've just updated the type of checks on the if statements - they now check `typeof === 'boolean'` instead of the undefined check before - however my comment in reply to yours (above) still stands :smile: (as the flow hasn't changed)

File: src/angular-local-storage.js [hunk 2]
```diff
@@ -318,9 +320,23 @@ angular
...
+               * (allows developer to mix and match secure + non-secure) */
+              if (typeof secure === 'boolean') {
+                  if (secure === true) {
+                      /* We've explicitly specified secure,
+                       * add the secure attribute to the cookie (after domain) */
+                      cookieDomain += "; secure";
+                  }
+                  // else - secure has been supplied but isn't true - so don't set secure flag, regardless of what config says
+              }
+              else if (typeof secure === 'boolean' && cookie.secure === true) {
```
Comment (ealves-pt) [inline src/angular-local-storage.js]: The `typeof secure === 'boolean'` should not be needed here. We reach this `else if` if `secure` is not a `boolean`. Besides this your explanation makes total sense. Thanks for the input :+1:

Comment (ry8806) [inline src/angular-local-storage.js]: no worries, thanks for helping getting this in!

**Comment by ry8806** [comment]
hi, just wondering if you've had a chance to look at getting this in yet?

**Comment by grevory** [comment]
I see a syntax error on line 98 in localStorageSpec.js.

**Comment by ry8806** [comment]
I've corrected the syntax error - apologies!

**Comment by grevory** [comment]
It looks like it is out of sync with master. Can you resolve this first? Thanks.

**Comment by grevory** [comment]
I don't want your hard work to go unused. Think you could correct the conflict with master?

**Comment by ry8806** [comment]
yes, I can have a look and see what I can do!

**Comment by ry8806** [comment]
it seems quite far behind, what do you reckon, just pull from master and re-do my changes on top, or rebase???

**Comment by jonkoops** [comment]
@ry8806 I would squash your commits into one and rebase from master.

**Comment by ry8806** [comment]
@jonkoops help would be much appreciated, only familiar with the (very) basic git stuff.
no idea how to squash and then rebase before all hell breaks loose on my machine! 😄

**Comment by jonkoops** [comment]
@ry8806 It seems there are some commits in your pull request which are not part of the functionality itself. Is this correct?

**Comment by jonkoops** [comment]
Looks like you pulled in some changes from the upstream master which are now part of your pull request. Could you give me access to your repository so I can correct it? I would love to explain all the inner workings of rebasing but I unfortunately do not have the time.

**Comment by ry8806** [comment]
think you should have access to the repo, no worries, thanks for helping!

i'll definitely have to read up more on the inner workings of squash/rebase when I find time

**Comment by jonkoops** [comment]
@ry8806 I've squashed all the changes into a single commit, and I've rebased with the upstream master. We should be all good to merge now (if the tests run correctly).

**Comment by ry8806** [comment]
massive thanks @jonkoops

**Comment by jonkoops** [comment]
Looks like some of the tests are failing, did I acidentally break something or were those tests failing before too?

**Comment by ry8806** [comment]
not sure, i'll check and fix 😄

**Comment by jonkoops** [comment]
Ok, I've rebased your changes into the base commit. With the tests passing we should be good, for realsies this time.

**Comment by jonkoops** [comment]
@grevory Any chance you could review and possibly merge?

**Comment by jonkoops** [comment]
@grevory Bump.

bump

**Comment by grevory** [comment]
Sorry. Been travelling for work and pleasure. Thanks a ton!

**Pull request merged**